### PR TITLE
TensorFlow: Fixed shape issue where unknown shape will be induced from data.

### LINKED ISSUE #2458: TensorFlowTransformer is not properly matching the input dimensions
### System information

- **OS version/distro**: Windows 10/Pro
- **.NET Version (eg., dotnet --info)**: .NET Version 2.1.202
- ML.TensorFlow Version 0.10

### Issue

- **What did you do?**: Executed a TensorFlow model using TensorFlowEstimator
- **What happened?**: Mapper does not properly match the input dimensions to the TensorFlow model dimensions
- **What did you expect?**: Properly matched dimensions

### Source code / logs

TensorFlowTransform.cs, Lines 810 - 875

Line 853:

var d = originalShape.NumDimensions > 2 ? Math.Pow(typeValueCount / valCount, 1.0 / (originalShape.NumDimensions - 2)) : 1;

The case with 2 dimensions, where the originalShape is [1, ?] is not matched properly.

### PR DESCRIPTION
**Pull request by zeahmed**
This PR fixes #2458.

The problem addressed in this PR are
1. Input is used to determine the shape. If there is one unknown dimension, it is computed from the input data. If there are several unknown dimension, the value computed from the input is equally distributed among unknown dimension.
2. There is no batch dimension assumption on input now.
3. Output shape are computed as previously done. However, if there are more than one unknown dimensions in the shape, a variable length vector is returned then.
4. Created test to check for different type of shapes can be pass in and out of TF.

### CODE REVIEW

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 1]
```diff
@@ -74,6 +74,103 @@ public void TensorFlowTransformMatrixMultiplicationTest()
...
+                                            FourDim = new[]{ 121.0f, 122.0f, 123.0f, 124.0f, 125.0f, 126.0f,
+                                                             127.0f, 128.0f, 129.0f, 130.0f, 131.0f, 132.0f},
+                                            FourDimKnown = new[]{ 141.0f , 142.0f, 143.0f, 144.0f, 145.0f, 146.0f, 147.0f, 148.0f,
+                                                                  149.0f , 150.0f, 151.0f, 152.0f, 153.0f, 154.0f, 155.0f, 156.0f }
+                                        }
+                });
+        }
+
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // TensorFlow is 64-bit only
+        public void TensorFlowTransformInputShapeTest()
```
Comment (zeahmed) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: Test is going to fail as the nuget is not updated yet.

File: src/Microsoft.ML.TensorFlow/TensorflowTransform.cs [hunk 2]
```diff
@@ -698,7 +689,14 @@ internal static (TFDataType[] tfOutputTypes, ColumnType[] outputTypes) GetOutput
             {
                 var tfOutput = new TFOutput(session.Graph[outputs[i]]);
                 var shape = session.Graph.GetTensorShape(tfOutput);
+
+                // The transformer can only retreive the output as fixed length vector with shape of kind [-1, d1, d2, d3, ...]
+                // i.e. the first dimension (if unknown) is assumed to be batch dimension.
+                // If there are other dimension that are unknown the transfomer will return a variable length vector.
+                // This is the work around in absence of reshape transofmer.
```
Comment (zeahmed) [inline src/Microsoft.ML.TensorFlow/TensorflowTransform.cs]: typo: transofmer -> transformer

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 1 above)
Comment (eerhardt) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: CI passed.  I assume the nuget has been updated?

---
In reply to: [254931624](https://github.com/dotnet/machinelearning/pull/2475#discussion_r254931624) [](ancestors = 254931624)

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 3]
```diff
@@ -74,6 +74,103 @@ public void TensorFlowTransformMatrixMultiplicationTest()
...
+        {
+            var modelLocation = "model_shape_test";
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: I assume `model_shape_test` has both unknown input and output dimensions, correct?

**Comment by wschin** [comment]
For (1), is it consistent to TF convention?

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 4]
```diff
@@ -74,6 +74,103 @@ public void TensorFlowTransformMatrixMultiplicationTest()
...
+            // Intended data shape [1, 2, 2], model shape [1, None, 2]
+            [VectorType(1, 2, 2)]
+            public float[] ThreeDim;
+
+            // Data will be passed as flat vector.
+            // Intended data shape [1, 2, 2, 3], model shape [1, None, None, 3]
+            [VectorType(12)]
+            public float[] FourDim;
+
+            // Data will be passed as 3-D vector.
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: >3 [](start = 38, length = 1)

4

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 5]
```diff
@@ -74,6 +74,103 @@ public void TensorFlowTransformMatrixMultiplicationTest()
             }
         }
 
+        private class ShapeData
+        {
+            // Data will be passed as flat vector.
+            // Intended data shape [2,2], model shape [2, None]
+            [VectorType(4)]
+            public float[] TwoDim;
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: >TwoDim [](start = 27, length = 6)

Is it possible to have [?] model shape?
I mean one dimensional case, would we handle it correctly?

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 1 above)
Comment (zeahmed) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: Yes, it is...:)

---
In reply to: [255142810](https://github.com/dotnet/machinelearning/pull/2475#discussion_r255142810) [](ancestors = 255142810,254931624)

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 3 above)
Comment (zeahmed) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: Yes, I described it on line 77-98

---
In reply to: [255145486](https://github.com/dotnet/machinelearning/pull/2475#discussion_r255145486) [](ancestors = 255145486)

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs (hunk 5 above)
Comment (zeahmed) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: Good catch. I should actually add a test for that too.

---
In reply to: [255220988](https://github.com/dotnet/machinelearning/pull/2475#discussion_r255220988) [](ancestors = 255220988)

**Comment by zeahmed** [comment]
For the case of one unknown dimension, yes. If there are more than one unknown dimension than there is no solution. This is just a way to it other way is to make a reshape transformer in ML.NET which user can use to define their input shape which then be mapped to TF shapes easily.


---
In reply to: [461891682](https://github.com/dotnet/machinelearning/pull/2475#issuecomment-461891682) [](ancestors = 461891682)

**Pull request merged**